### PR TITLE
bnd-workspace-example

### PR DESCRIPTION
**Pull request by stbischof**
An Example of an minimal bnd-workspace using `osgi-test`
@kriegfrj What would be the best way to integrate this in the `osgi-tests-test-setup`?

### CODE REVIEW

File: examples/osgi-test-example-bnd/cnf/build.bnd [hunk 1]
```diff
@@ -0,0 +1,2 @@
+## Place your settings in ext/.
+cnf: ${workspace}/cnf
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/build.bnd:2]: Huh? There is already `${build}` macro which does this.

File: examples/osgi-test-example-bnd/cnf/ext/defaults.bnd [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+Test-Cases: ${classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.junit.platform.commons.annotation.Testable;CONCRETE}
+
+-tester: biz.aQute.tester.junit-platform
+
+
+-resolve.effective: active
+
+
+# define global blacklist
+-runblacklist.default: osgi.identity;filter:='(osgi.identity=osgi.cmpn)'
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/defaults.bnd:40]: Why is this needed? You should NOT have osgi.cmpn in the repository. Use the small API jars.

File: examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+		index=${cnf}/ext/central.maven; \
+		name="Maven Central"
+
+-plugin.Local: \
+	aQute.bnd.deployer.repository.LocalIndexedRepo; \
+		name = Local; \
+		pretty = true; \
+		local = ${build}/local
+
+-plugin.Temp: \
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd:14]: What is this repo for?

File: examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Configure Repositories
+-plugin.Central: \
+	aQute.bnd.repository.maven.provider.MavenBndRepository; \
+		releaseUrl=https://repo.maven.apache.org/maven2/; \
+		index=${cnf}/ext/central.maven; \
+		name="Maven Central"
+
+-plugin.Local: \
+	aQute.bnd.deployer.repository.LocalIndexedRepo; \
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd:9]: What is this repo for?

File: examples/osgi-test-example-bnd/.gradle-wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/.gradle-wrapper/gradle-wrapper.properties:1]: We don't want gradlew in here. We did not put mvnw in the maven example. This just makes the example too large and confusing.

File: examples/osgi-test-example-bnd/cnf/ext/central.maven [hunk 6]
```diff
@@ -0,0 +1,45 @@
+org.osgi:org.osgi.test.junit5:0.10.0
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/central.maven:1]: We should use macros for the version so we don't have to change `0.10.0` many times. Also use macros for the junit, assertj, bnd versions too.

See https://github.com/bndtools/bnd/blob/a00db24d6526bbacd2d81700de2c3c9cb8e5001a/cnf/ext/central.mvn#L70-L86 for example.

File: examples/osgi-test-example-bnd/cnf/ext/central.maven [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ch.qos.logback:logback-core:1.2.3
+org.slf4j:jul-to-slf4j:1.7.25
+org.slf4j:log4j-over-slf4j:1.7.25
+org.slf4j:osgi-over-slf4j:1.7.25
+org.slf4j:slf4j-api:1.7.25
+org.slf4j:slf4j-ext:1.7.25
+org.slf4j:slf4j-jcl:1.7.25
+
+org.osgi:osgi.annotation:7.0.0
+org.osgi:osgi.cmpn:7.0.0
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/central.maven:41]: No osgi.cmpn.

File: examples/osgi-test-example-bnd/cnf/ext/defaults.bnd [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+javac.target: 1.8
+
+# This is the version of JUnit that will be used at build time and runtime
+osgitest: org.osgi.test.junit5;version=latest,\
+	org.osgi.test.junit5.cm;version=latest,\
+	org.osgi.test.common;version=latest,\
+	org.osgi.test.assertj.framework;version=latest
+
+# This is the version of JUnit that will be used at build time and runtime
+junit: junit-jupiter-api;version="[5.7.0,6)",\
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/defaults.bnd:13]: Why version ranges here? Just use latest since you control the version in the repo.

File: examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# Configure Repositories
+-plugin.Central: \
+	aQute.bnd.repository.maven.provider.MavenBndRepository; \
+		releaseUrl=https://repo.maven.apache.org/maven2/; \
+		index=${cnf}/ext/central.maven; \
+		name="Maven Central"
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd:6]: `readOnly=true`

File: examples/osgi-test-example-bnd/itest/.settings/org.eclipse.core.resources.prefs [hunk 10]
```diff
@@ -0,0 +1,3 @@
+eclipse.preferences.version=1
+encoding//src/itest/ExampleIntegrationTest.java=UTF-8
+encoding/bnd.bnd=UTF-8
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/itest/.settings/org.eclipse.core.resources.prefs:3]: `encoding/<project>=UTF-8` covers the whole project not just individual files.

File: examples/osgi-test-example-bnd/cnf/.project [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+<projectDescription>
+	<name>cnf</name>
+	<comment></comment>
+	<projects>
+	</projects>
+	<buildSpec>
+	</buildSpec>
+	<natures>
+	</natures>
+</projectDescription>
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/.project:11]: Why does `cnf` have no `.settings/`?

File: examples/osgi-test-example-bnd/itest/bnd.bnd [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	bnd.identity;id='junit-jupiter-engine'
+
+# Build dependencies 
+-buildpath: \
+	${junit},\
+	${mockito},\
+	${osgitest},\
+	osgi.annotation;version='7.0',\
+	osgi.core;version='6.0',\
+	osgi.cmpn;version='7.0',\
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/itest/bnd.bnd:12]: NO osgi.cmpn

File: examples/osgi-test-example-bnd/itest/bnd.bnd [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	junit-jupiter-api;version='5.7.0'
+	
+javac.source: 11
+javac.target: 11
+
+# We need JUnit and Mockito to resolve the test cases at runtime. 
+# Other runtime dependencies should be added as necessary
+-runbundles: \
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/itest/bnd.bnd:20]: Sorting like we do in maven example:

```
# This will help us keep -runbundles sorted
-runstartlevel: \
    order=sortbynameversion,\
    begin=-1
```

File: examples/osgi-test-example-bnd/itest/bnd.bnd [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+	org.osgi.test.junit5.cm;version='[0.10.0,0.10.1)',\
+	itest;version=snapshot,\
+	junit-jupiter-engine;version='[5.7.0,5.7.1)'
+
+
+Bundle-Version: 0.0.0.${tstamp}
+Private-Package: itest
+
+# Needed for Mockito's mocking to work
+-runsystempackages.objenesis: sun.misc,sun.reflect
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/itest/bnd.bnd:38]: This should be in cnf near the definition of the mockito macro.

File: examples/osgi-test-example-bnd/itest/bnd.bnd [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+# Use Felix by default
+-runfw: org.apache.felix.framework;version='[6.0.1,6.0.1]'
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/itest/bnd.bnd:41]: Why hard code version when only one version exists in the repo?

File: examples/osgi-test-example-bnd/itest/.project [hunk 16]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>itest</name>
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/itest/.project:3]: Can we use a more meaningful project name than `itest`? See the maven example.

**Review by bjhargrave — CHANGES_REQUESTED** [review]
This needs a lot of work.

Why is this example not the same as the maven example but adapted for Bnd workspace build?

File: examples/osgi-test-example-bnd/itest/bnd.bnd (hunk 15 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/itest/bnd.bnd]: removed version

File: examples/osgi-test-example-bnd/itest/bnd.bnd (hunk 14 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/itest/bnd.bnd]: moved to cnd/defaults.bnd

File: examples/osgi-test-example-bnd/itest/bnd.bnd (hunk 13 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/itest/bnd.bnd:24]: inserted

File: examples/osgi-test-example-bnd/itest/bnd.bnd (hunk 12 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/itest/bnd.bnd]: removed osgi.cmpn;

File: examples/osgi-test-example-bnd/cnf/.project (hunk 11 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/.project:11]: added minimum .settings

File: examples/osgi-test-example-bnd/itest/.settings/org.eclipse.core.resources.prefs (hunk 10 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/itest/.settings/org.eclipse.core.resources.prefs]: thx. done

File: examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd (hunk 9 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd:7]: thx. done

File: examples/osgi-test-example-bnd/cnf/ext/defaults.bnd (hunk 8 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/defaults.bnd]: changes to latest

File: examples/osgi-test-example-bnd/cnf/ext/central.maven (hunk 7 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/central.maven]: removed osgi.cmpn;

File: examples/osgi-test-example-bnd/.gradle-wrapper/gradle-wrapper.properties (hunk 5 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/.gradle-wrapper/gradle-wrapper.properties]: removed gradle

File: examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd (hunk 4 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd]: dont know. removed

File: examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd (hunk 3 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/repo_common.bnd]: dont know, removed

File: examples/osgi-test-example-bnd/cnf/ext/defaults.bnd (hunk 2 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/defaults.bnd]: removed osgi.cmpn

File: examples/osgi-test-example-bnd/cnf/build.bnd (hunk 1 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/build.bnd]: removed

**Comment by stbischof** [comment]
I am struggeling a bit here. searching for a way where i can have the bundle together with the tests-bundle(as fragment) in the same project ( like in maven)

**Comment by kriegfrj** [comment]
You could use sub bundles and have one for the main bundle (containing the application code) and the other for the fragment containing the tests. All code will go in the src dir, not the test dir.

**Comment by bjhargrave** [comment]
> I am struggeling a bit here. searching for a way where i can have the bundle together with the tests-bundle(as fragment) in the same project ( like in maven)

But this is probably a bad idea. Since your test bundle will need test code (like Junit) visible in the `-buildpath` since you need to build the test bundle from things in the `-buildpath`, you then expose your main bundle to inadvertently referencing test code in error. So for Bnd Workspace builds, OSGi test bundle should be in their own project separate from any bundles which they test.

The maven example can build the test bundle from the test path while the main bundle is built from the compile path. This is not possible in Bnd Workspace builds.

**Comment by bjhargrave** [comment]
> All code will go in the src dir, not the test dir.

But this is a bad "example" to put forward for others to use. The point of theses examples is to show the proper was to do things and not an improper way.

**Comment by kriegfrj** [comment]
> But this is a bad "example" to put forward for others to use. The point of theses examples is to show the proper was to do things and not an improper way.

Agreed. 

Wasn't the original example i set up (on which the Maven example is based) set up that way? Can we just use that? Is Stefan reinventing the wheel here?

**Comment by stbischof** [comment]
just to be clear.  I would provide the example a user would expect.  

- [ ]  one Project for the normal bundle
- [ ]   one Project for the itest bundle
- [ ]  one `extra` Project witch shoes how to build a fooder bundle with `-make` and test it wit InstallBundle
- [ ]   one `extra` AllInOne Project which holds sources and test in seperate test folders and creates a tests-fragment and the the -testpath is only visible in the tests

**Comment by bjhargrave** [comment]
>  But also want to add extra projects that shows more spectial ways that i would use.

This seems not a good idea for an example of how to do things. Showing unusual or atypical ways of doing things here would just be confusing.

**Comment by stbischof** [comment]
does not work at the moment.
running the bndrun as osgi-test results in an

```
 Unexpected error in the run body: null
java.lang.NullPointerException
	at java.base/java.util.StringTokenizer.<init>(StringTokenizer.java:199)
	at java.base/java.util.StringTokenizer.<init>(StringTokenizer.java:221)
	at aQute.launcher.constants.LauncherConstants.split(LauncherConstants.java:159)
	at aQute.launcher.constants.LauncherConstants.<init>(LauncherConstants.java:137)
	at aQute.launcher.Launcher.launch(Launcher.java:392)
	at aQute.launcher.Launcher.run(Launcher.java:185)
	at aQute.launcher.Launcher.main(Launcher.java:161)
	at aQute.launcher.pre.EmbeddedLauncher.executeWithRunPath(EmbeddedLauncher.java:170)
	at aQute.launcher.pre.EmbeddedLauncher.findAndExecute(EmbeddedLauncher.java:135)
	at aQute.launcher.pre.EmbeddedLauncher.main(EmbeddedLauncher.java:52)
java.lang.NullPointerException
	at aQute.launcher.Launcher.trace(Launcher.java:1562)
	at aQute.launcher.Launcher.launch(Launcher.java:460)
	at aQute.launcher.Launcher.run(Launcher.java:185)
	at aQute.launcher.Launcher.main(Launcher.java:161)
	at aQute.launcher.pre.EmbeddedLauncher.executeWithRunPath(EmbeddedLauncher.java:170)
	at aQute.launcher.pre.EmbeddedLauncher.findAndExecute(EmbeddedLauncher.java:135)
	at aQute.launcher.pre.EmbeddedLauncher.main(EmbeddedLauncher.java:52)
```

**Comment by bjhargrave** [comment]
> running the bndrun as osgi-test results in an
> 
> ```
> 
> ```

This is a bug in 5.4.0-SNAPSHOT which has been fixed.

**Comment by stbischof** [comment]
@bjhargrave  Is there more to do do to get this passed?

File: examples/osgi-test-example-bnd/cnf/ext/central.maven [hunk 17]
```diff
@@ -0,0 +1,49 @@
+org.osgi:org.osgi.test.junit5:0.10.0
+org.osgi:org.osgi.test.junit5.cm:0.10.0
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/central.maven:2]: These need to be updated to 1.0.0

File: examples/osgi-test-example-bnd/cnf/ext/central.maven [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+org.osgi:org.osgi.test.common:0.10.0
+org.osgi:org.osgi.test.assertj.framework:0.10.0
+org.assertj:assertj-core:3.18.0
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/cnf/ext/central.maven:5]: Need to update the versions of assertj, mockito, Junit, bnd, etc. All these versions are somewhat old now.

File: examples/osgi-test-example-bnd/gradle.properties [hunk 19]
```diff
@@ -0,0 +1,2 @@
+bnd_version=5.2.0
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/gradle.properties:1]: 5.3.0

File: examples/osgi-test-example-bnd/org.osgi.test.example.api/bnd.bnd [hunk 20]
```diff
@@ -0,0 +1,4 @@
+# Build dependencies 
+-buildpath: osgi.annotation
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.api/bnd.bnd:2]: Should have version attribute.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/.settings/org.eclipse.m2e.core.prefs [hunk 21]
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/.settings/org.eclipse.m2e.core.prefs:1]: Remove this prefs file since this is not a maven example.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/.settings/org.eclipse.m2e.core.prefs [hunk 22]
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.test/.settings/org.eclipse.m2e.core.prefs:1]: Remove. This is not a maven example.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd [hunk 23]
```diff
@@ -0,0 +1,7 @@
+# Build dependencies 
+-buildpath: \
+	osgi.annotation,\
+	org.osgi.test.example.api;version=latest,\
+	osgi.cmpn
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd:5]: Don't use osgi.cmpn. It is now an anti-practice and we don't want to show using it in an example. You need to use the individual spec API jars as needed.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd [hunk 24]
```diff
@@ -0,0 +1,7 @@
+# Build dependencies 
+-buildpath: \
+	osgi.annotation,\
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd:3]: Need version attribute.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd [hunk 25]
```diff
@@ -0,0 +1,17 @@
+# Build dependencies 
+-buildpath: \
+	${junit},\
+	${mockito},\
+	${osgitest},\
+	osgi.annotation;version='7.0',\
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd:6]: Why specific versions here instead of latest? The buildpaths across the projects should have a consistent look.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+	osgi.core;version='6.0',\
+	junit-jupiter-api;version='5.7.0',\
+	org.osgi.test.example.api,\
+	osgi.cmpn,\
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd:10]: No osgi.cmpn

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+	org.apiguardian:apiguardian-api
+
+
+Bundle-Version: 0.0.0.${tstamp}
+
+# This would execute the bndrun while Gradle build
+-include:  test.bndrun
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd:17]: This looks like a bad practice. One should not include a bndrun into a bnd.bnd.

**Review by bjhargrave — CHANGES_REQUESTED** [review]

**Comment by bjhargrave** [comment]
I would also like the directory name for the example to be ` examples/osgi-test-example-bnd-workspace` to make it clear this is a Bnd Workspace example rather than just a Bnd example since all the examples use Bnd plugins.

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd (hunk 26 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd]: removed cmpn also from the repo file

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd (hunk 25 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd]: use lestest in all bnd.bnd files

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd (hunk 24 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd]: use latest like in als bnd files

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd (hunk 23 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/bnd.bnd]: removed cmpn also from the repo file

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/.settings/org.eclipse.m2e.core.prefs (hunk 22 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.test/.settings/org.eclipse.m2e.core.prefs]: reoved

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/.settings/org.eclipse.m2e.core.prefs (hunk 21 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.impl/.settings/org.eclipse.m2e.core.prefs]: removed

File: examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd (hunk 27 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.player.test/bnd.bnd]: i put it into the bnd file

File: examples/osgi-test-example-bnd/org.osgi.test.example.api/bnd.bnd (hunk 20 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/org.osgi.test.example.api/bnd.bnd]: ok

File: examples/osgi-test-example-bnd/gradle.properties (hunk 19 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/gradle.properties]: ok

File: examples/osgi-test-example-bnd/cnf/ext/central.maven (hunk 18 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/central.maven]: updated all versions

File: examples/osgi-test-example-bnd/cnf/ext/central.maven (hunk 17 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd/cnf/ext/central.maven]: updated all versions

File: examples/osgi-test-example-bnd-workspace/cnf/eclipse/jdt.bnd [hunk 28]
```diff
@@ -0,0 +1,4 @@
+-include: ${project}/.settings/org.eclipse.jdt.core.prefs
+javac.source 			= ${org.eclipse.jdt.core.compiler.source}
+javac.target 			= ${org.eclipse.jdt.core.compiler.codegen.targetPlatform}
+javac.compliance 		= ${org.eclipse.jdt.core.compiler.compliance}
```
[the remark this answers is not in the document]
Comment (bjhargrave) [reply examples/osgi-test-example-bnd-workspace/cnf/eclipse/jdt.bnd:4]: I don't see the `org.eclipse.jdt.core.prefs` files listing the source and target values?

File: examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.impl/bnd.bnd [hunk 29]
```diff
@@ -0,0 +1,7 @@
+# Build dependencies 
+-buildpath: \
+	osgi.annotation;version=latest,\
+	org.osgi.test.example.api;version=latest,\
+	org.osgi.service.component.annotations;version=latest
+
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.impl/bnd.bnd:6]: I think you need  -testpath for the junit test that should be in this project. Where is the OpenBox test class?

File: examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/.settings/org.eclipse.core.resources.prefs [hunk 30]
```diff
@@ -0,0 +1,3 @@
+eclipse.preferences.version=1
+encoding//src/main/java=UTF-8
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/.settings/org.eclipse.core.resources.prefs:2]: remove

File: examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/bnd.bnd [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+-runpath: \
+	ch.qos.logback.classic,\
+	ch.qos.logback.core,\
+	org.apache.felix.logback,\
+	slf4j.api
+	
+-runrequires: \
+	bnd.identity;id='junit-jupiter-engine',\
+	bnd.identity;id='junit-platform-launcher',\
+	bnd.identity;id='org.osgi.test.example.player.test'
```
Comment (bjhargrave) [inline examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/bnd.bnd:39]: This should use a macro like the other examples. `id=${p}`

File: examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.project [hunk 32]
```diff
@@ -11,13 +11,13 @@
 			</arguments>
 		</buildCommand>
 		<buildCommand>
-			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<name>bndtools.core.bndbuilder</name>
```
Comment (bjhargrave) [inline examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.project:14]: Why are you breaking the osgi-test-example-mvn project Here?

File: examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.settings/org.eclipse.jdt.core.prefs [hunk 33]
```diff
@@ -165,7 +165,6 @@ org.eclipse.jdt.core.formatter.alignment_for_arguments_in_explicit_constructor_c
 org.eclipse.jdt.core.formatter.alignment_for_arguments_in_method_invocation=16
 org.eclipse.jdt.core.formatter.alignment_for_arguments_in_qualified_allocation_expression=16
 org.eclipse.jdt.core.formatter.alignment_for_assignment=0
-org.eclipse.jdt.core.formatter.alignment_for_binary_expression=16
 org.eclipse.jdt.core.formatter.alignment_for_bitwise_operator=16
```
Comment (bjhargrave) [inline examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.settings/org.eclipse.jdt.core.prefs:168]: This is in the osgi-test-example-mvn project.

**Review by bjhargrave — CHANGES_REQUESTED** [review]

File: examples/osgi-test-example-bnd-workspace/cnf/ext/defaults.bnd [hunk 34]
```diff
@@ -0,0 +1,33 @@
...
+	assertj-core;version=latest
+
+# This is the version of Mockito that will be used at build time and run time
+mockito: org.mockito.mockito-core;version=latest,\
+	org.objenesis;version=latest
+	
+# Needed for Mockito's mocking to work
+-runsystempackages.objenesis: sun.misc,sun.reflect
+
+Test-Cases: ${classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.junit.platform.commons.annotation.Testable;CONCRETE}
```
[the remark this answers is not in the document]
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/cnf/ext/defaults.bnd:29]: bnd will only set this header if it is not empty. thought this is a feature

File: examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.project (hunk 32 above)
Comment (stbischof) [reply examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.project:14]: used a wrong git add filter

File: examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.settings/org.eclipse.jdt.core.prefs (hunk 33 above)
Comment (stbischof) [reply examples/osgi-test-example-mvn/org.osgi.test.example.player.impl/.settings/org.eclipse.jdt.core.prefs:168]: used a wrong git add filter

File: examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/bnd.bnd (hunk 31 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/bnd.bnd:39]: now its `bnd.identity;id='${p}.test'`

File: examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/.settings/org.eclipse.core.resources.prefs (hunk 30 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.test/.settings/org.eclipse.core.resources.prefs:2]: ok

File: examples/osgi-test-example-bnd-workspace/gradle.properties [hunk 35]
```diff
@@ -0,0 +1,2 @@
+bnd_version=5.2.0
```
[the remark this answers is not in the document]
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/gradle.properties:1]: changed

File: examples/osgi-test-example-bnd-workspace/cnf/ext/defaults.bnd [hunk 36]
```diff
@@ -0,0 +1,33 @@
+java: java
+javac: javac
+javac.source: 11
+javac.target: 11
+
+# This is the version of JUnit that will be used at build time and runtime
```
[the remark this answers is not in the document]
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/cnf/ext/defaults.bnd:6]: renamed.

File: examples/osgi-test-example-bnd-workspace/cnf/eclipse/jdt.bnd (hunk 28 above)
[the remark this answers is not in the document]
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/cnf/eclipse/jdt.bnd:4]: added `-include: ${workspace}/cnf/eclipse/jdt.bnd`to every project bnd

File: examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.impl/bnd.bnd (hunk 29 above)
Comment (stbischof) [reply examples/osgi-test-example-bnd-workspace/org.osgi.test.example.player.impl/bnd.bnd:12]: added `OpenBox.java` from mvn

**Comment by stbischof** [comment]
Pull request successfully merged and closed ?

**Comment by bjhargrave** [comment]
Please see https://github.com/osgi/osgi-test/pull/363 for my alternative to this PR.
